src/libpriv/rpmostree-rpm-util.cxx

```diff
@@ -1266,6 +1266,30 @@ rpmostree_create_rpmdb_pkglist_variant (int dfd, const char *path, GVariant **ou
   return TRUE;
 }
 
+namespace rpmostreecxx
+{
+rust::Vec<rust::String>
```

Comment:
Now that you've improved the DnfPackage bindings I think we could use that for this in the future, right?

Reply:
Ideally yes, but I'm not sure how to return a collection of opaque C++ types. There's `CxxVector`, but it doesn't seem supported in return position. There's `CxxGObjectArray` but that would require unsafe code to use.

Reply:
OK, this is how far I got:

```diff
diff --git a/rust/src/fedora_integration.rs b/rust/src/fedora_integration.rs
index dbc48925..4179ce86 100644
--- a/rust/src/fedora_integration.rs
+++ b/rust/src/fedora_integration.rs
@@ -223,6 +223,7 @@ fn get_base_package_list() -> Result<HashSet<String>> {
         Ok(
             crate::ffi::rpmdb_package_name_list(libc::AT_FDCWD, "/".to_string())?
                 .into_iter()
+                .map(|pkg| pkg.get_name())
                 .collect(),
         )
     } else {
diff --git a/rust/src/lib.rs b/rust/src/lib.rs
index ff1aeaa8..c679f60b 100644
--- a/rust/src/lib.rs
+++ b/rust/src/lib.rs
@@ -62,6 +62,8 @@ pub mod ffi {

         #[namespace = "dnfcxx"]
         type FFIDnfPackage = libdnf_sys::FFIDnfPackage;
+        #[namespace = "dnfcxx"]
+        type DnfPackage = libdnf_sys::DnfPackage;
     }

     /// Currently cxx-rs doesn't support mappings; like probably most projects,
@@ -821,7 +823,7 @@ pub mod ffi {
         fn nevra_to_cache_branch(nevra: &CxxString) -> Result<String>;
         fn get_repodata_chksum_repr(pkg: &mut FFIDnfPackage) -> Result<String>;
         fn rpmts_for_commit(repo: Pin<&mut OstreeRepo>, rev: &str) -> Result<UniquePtr<RpmTs>>;
-        fn rpmdb_package_name_list(dfd: i32, path: String) -> Result<Vec<String>>;
+        fn rpmdb_package_name_list(dfd: i32, path: String) -> UniquePtr<CxxVector<DnfPackage>>;

         // Methods on RpmTs
         fn packages_providing_file(self: &RpmTs, path: &str) -> Result<Vec<String>>;
diff --git a/src/libpriv/rpmostree-rpm-util.cxx b/src/libpriv/rpmostree-rpm-util.cxx
index 90b76b76..f6f7bc50 100644
--- a/src/libpriv/rpmostree-rpm-util.cxx
+++ b/src/libpriv/rpmostree-rpm-util.cxx
@@ -1268,7 +1268,7 @@ rpmostree_create_rpmdb_pkglist_variant (int dfd, const char *path, GVariant **ou

 namespace rpmostreecxx
 {
-rust::Vec<rust::String>
+std::unique_ptr<std::vector<dnfcxx::DnfPackage>>
 rpmdb_package_name_list (gint32 dfd, rust::String path)
 {
   g_autoptr (GError) local_error = NULL;
@@ -1277,16 +1277,16 @@ rpmdb_package_name_list (gint32 dfd, rust::String path)
   if (!refsack)
     throw std::runtime_error (local_error->message);

-  rust::Vec<rust::String> r;
+  std::vector<dnfcxx::DnfPackage> r;
   g_autoptr (GPtrArray) pkglist = rpmostree_sack_get_sorted_packages (refsack->sack);
   const guint n = pkglist->len;
   for (guint i = 0; i < n; i++)
     {
       auto pkg = static_cast<DnfPackage *> (pkglist->pdata[i]);
-      r.push_back (rust::String (dnf_package_get_name (pkg)));
+      r.push_back (dnfcxx::DnfPackage ((DnfPackage*)g_object_ref (pkg)));
     }

-  return r;
+  return std::make_unique<std::vector<dnfcxx::DnfPackage>> (r);
 }
 }

diff --git a/src/libpriv/rpmostree-rpm-util.h b/src/libpriv/rpmostree-rpm-util.h
index 9e848cf2..a6f4de7e 100644
--- a/src/libpriv/rpmostree-rpm-util.h
+++ b/src/libpriv/rpmostree-rpm-util.h
@@ -33,13 +33,15 @@

 #include "libglnx.h"

+#include "rust/libdnf-sys/cxx/libdnf.hpp"
+
 // C++ code
 namespace rpmostreecxx
 {
 rust::String nevra_to_cache_branch (const std::string &nevra);
 rust::String get_repodata_chksum_repr (DnfPackage &pkg);
 std::unique_ptr<RpmTs> rpmts_for_commit (OstreeRepo &repo, rust::Str rev);
-rust::Vec<rust::String> rpmdb_package_name_list (gint32 dfd, rust::String path);
+std::unique_ptr<std::vector<dnfcxx::DnfPackage>> rpmdb_package_name_list (gint32 dfd, rust::String path);
 }

 // C code follows
```

But I'm hitting

```
error[E0277]: the trait bound `libdnf_sys::DnfPackage: VectorElement` is not satisfied
   --> rust/src/lib.rs:826:63
    |
826 |         fn rpmdb_package_name_list(dfd: i32, path: String) -> UniquePtr<CxxVector<DnfPackage>>;
    |                                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `VectorElement` is not implemented for `libdnf_sys::DnfPackage`
    |
    = note: required because of the requirements on the impl of `UniquePtrTarget` for `CxxVector<libdnf_sys::DnfPackage>`
note: required by a bound in `UniquePtr`
   --> /var/home/jlebon/.cargo/registry/src/github.com-1ecc6299db9ec823/cxx-1.0.68/src/unique_ptr.rs:17:8
    |
17  |     T: UniquePtrTarget,
    |        ^^^^^^^^^^^^^^^ required by this bound in `UniquePtr`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `rpmostree-rust` due to previous error
make[2]: *** [Makefile:4759: cargo-build] Error 101
make[2]: Leaving directory '/var/home/jlebon/Code/github.com/coreos/rpm-ostree'
make[1]: *** [Makefile:3779: all-recursive] Error 1
make[1]: Leaving directory '/var/home/jlebon/Code/github.com/coreos/rpm-ostree'
make: *** [Makefile:1635: all] Error 2
```

https://docs.rs/cxx/latest/cxx/vector/trait.VectorElement.html says:

> This trait has no publicly callable or implementable methods. Implementing it outside of the CXX codebase is not supported.